core/state/statedb_logger.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/stateless"
+	"github.com/ethereum/go-ethereum/core/tracing"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie/utils"
+	"github.com/holiman/uint256"
+)
+
+type stateDBLogger struct {
+	inner  *StateDB
+	logger *tracing.Hooks
+}
+
+func newStateDBLogger(db *StateDB, logger *tracing.Hooks) *stateDBLogger {
+	s := &stateDBLogger{db, logger}
+
+	db.SetBurnCallback(func(address common.Address, amount *uint256.Int) {
+		if s.logger != nil && s.logger.OnBalanceChange != nil {
+			s.logger.OnBalanceChange(address, amount.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestructBurn)
+		}
+	})
+	return s
+}
+
+func (s *stateDBLogger) CreateAccount(address common.Address) {
+	s.inner.CreateAccount(address)
+}
+
+func (s *stateDBLogger) CreateContract(address common.Address) {
+	s.inner.CreateContract(address)
+}
+
+func (s *stateDBLogger) SubBalance(address common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
+	return s.inner.SubBalance(address, amount, reason)
+}
+
+func (s *stateDBLogger) AddBalance(address common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
+	prev := s.inner.AddBalance(address, amount, reason)
+	if s.logger != nil && s.logger.OnBalanceChange != nil {
```

Comment:
We can probably drop the s.logger != nil check in this object. I think it is created only in case there is a logger?

Reply:
As it is right now, no, because it's set unconditionally in blockchain.go, might still be nil. But it probably makes sense to not even use this layer in case it is nil, so you are right that we should remove this check eventually